### PR TITLE
[WIP] major change! reducing GLAbstraction to the basics

### PR DESCRIPTION
**Pull request by SimonDanisch**

### CODE REVIEW

**Comment by SimonDanisch** [comment]
Immediate goal: Make RenderObject usable again - I completely destroyed the functionality of it in this refactor.

General Goals:

* getting rid of Reactive/Color/ and other not strictly opengl related packages. Instead offer overloadable APIs to do the job

* Introduce leaner VertexArray buffer, integrating nicely with `view(buffer, faces)`. A mesh is then basically just `view(vertices::Vector{Point3f0}, indices::Vector{GLTriangle})`

* Introduce UniformBuffers to hold state in shaders independent of executing the shader

* Introduce lean RenderObject, that doesn't hold any data, besides information on the shader layout - data will get transferred via calling the object with new data. When uniformbuffers are used, data can also be updated in place

* remove weird GLVisualize code, that was basically just parked here because I didn't had a better place to put it

* Transpiler integration - make it the main way to create shaders, instead of having ugly templated shader that nobody understands

File: src/for_moderngl.jl [hunk 1]
```diff
@@ -229,3 +226,26 @@ function glTexImage(ttype::GLenum, level::Integer, internalFormat::GLenum, w::In
     end
     glTexImage1D(ttype, level, internalFormat, w, border, format, datatype, data)
 end
+
+function compile_program(shaders::GLuint...)
+    program = createprogram()
```
[the remark this answers is not in the document]
Comment (louisponet) [reply src/for_moderngl.jl]: Oh, this is not used anymore in the creation of a Program, It's done in the constructor now, there should be error handling there.

File: src/buffer.jl [hunk 2]
```diff
@@ -1,66 +1,72 @@
-mutable struct GLBuffer{T} <: GPUArray{T, 1}
+mutable struct Buffer{T} <: GPUArray{T, 1}
     id          ::GLuint
-    size        ::NTuple{1, Int}
+    size        ::Int
```
[the remark this answers is not in the document]
Comment (louisponet) [reply src/buffer.jl]: done

File: example/compute_shader.jl [hunk 3]
```diff
@@ -57,7 +57,7 @@ const roll = const_lift(every(0.1)) do x
     i::Float32 += 0.01f0
 end
 const tex = Texture(Float32, (512, 512))
-glBindImageTexture(0, tex.id, 0, GL_FALSE, 0, GL_WRITE_ONLY, tex.internalformat);
+glBindImageTexture(0, tex.GLid, 0, GL_FALSE, 0, GL_WRITE_ONLY, tex.internalformat);
```
[the remark this answers is not in the document]
Comment (louisponet) [reply example/compute_shader.jl]: yea was changed back again, was silly sorry

File: src/for_moderngl.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (louisponet) [reply src/for_moderngl.jl]: added the error checking into compile_program too

**Comment by karlwessel** [comment]
I'm not sure if this is the best place but it seems like the most active/promising part of GLAbstraction to me to ask:
What is the state of this rewrite and GLAbstraction for Julia 1.x in general?

As context:
I'm currently developing a lot of OpenGL related code in Julia and would love to use a more julian abstraction layer to OpenGL instead of having to write everything with ModernGL myself (which ends up looking like most of the functionality already present in GLAbstraction anyway). 

I know that there is (GL)Makie but that requires OpenGL 3.0, which is not available to me (ModernGL works fine with my OpenGL 2.1 hardware). It also seems odd to me, having to import GLMakie and therefore the whole AbstractPlotting backend when I only need the GLAbstraction layer of it.

I have seen that @louisponet seems to be actively developing his [own branch](https://github.com/louisponet/GLAbstraction.jl/tree/v1.0%2Brefactor) of GLAbstraction which makes me hope that we could have a clean and Julia 1.x compatible abstraction layer at some sooner point in the future.

What needs to be done? Do you need some help? I would be especially willing to write tests and fix bugs.

**Comment by SimonDanisch** [comment]
Is there anything in the way of using [Makie](https://github.com/JuliaPlots/Makie.jl/) & [GLMakie](https://github.com/JuliaPlots/GLMakie.jl)? You can actually extend GLMakie with your own shaders, if anything is missing! I incorporated a 1.0 compatible GLAbstraction module into GLMakie, so you could also use that layer - although I'm not really sure why anyone would want to miss out on the features of Makie, when using OpenGL ;)

**Comment by karlwessel** [comment]
I would love to use Makie, but since I only have OpenGL 2.1 compatible hardware it does not even build. I even started removing the hard check for 3.0 in GLMakies _deps/build.jl_ but stumpled from one incompatibility to the next and at some point I thought "well, maybe the hard check for OpenGL 3.0 is there for a good reason" :-).

Anyway it would be way easier for me to make only GLAbstraction (especially the reduced version from this PR) OpenGL 2.1 compatible than the full GLMakie.

**Comment by SimonDanisch** [comment]
Ohh, how did you end up with OpenGL 2.1? Are you sure that this isn't just a driver problem? There are very few GPUs, that don't support 3.1...

**Comment by karlwessel** [comment]
I'm using an up-to-date arch with the xf86-video-intel driver. It is a rather old Thinkpad T410s without a dedicated graphics card. It does support surprisingly many of the OpenGL 3+ features by extensions but some important things like GLSL > 1.2 syntax obviously not.

**Comment by louisponet** [comment]
Yea the branch I'm working with works perfectly with 1.0, I'm using it extensively in `Glimpse.jl`. What features are you looking for, all the basics should be there. There is no documentation however, if you have questions about things let me know. I'm focused on post 3.0 OpenGL thought so I'm not sure how well everything will work on lower versions. 

About it being clean, I feel like it is clean enough at least for what I use it :p, but I'm open to suggestions. From my point of view nothing really still needs to be done to get the core functionality working, but I for sure don't use all the features of OpenGL so your mileage may vary.

Cheers

**Comment by karlwessel** [comment]
> Yea the branch I'm working with works perfectly with 1.0, I'm using it extensively in `Glimpse.jl`. 

It is good news to hear that your branch is under active development and in use!

I started using the _v1.0+refactor_ branch from your repository yesterday by writing tests for the shader and program functionality that I would need and after that started replacing my own shader and programm abstractions with the ones from your branch. It already worked pretty good but there were some issues I had to fix beforehand:
- Using your default implementations of `gluniform` which use `glUniform1fv` does not work on my hardware when trying to set a `mat4` or `vec3` uniform from a 4x4 or 3x1 `Array` of `GLfloat`s. I had to implement a method for `gluniform` that explicitly checks the size of the array and uses the appropriate `glUniformMatrix4fv` and `glUniform3fv` function.
- 2.1 does not support the "`layout(location = 0)`" syntax in shaders, you have to define the location using `glBindAttribLocation` before linking the program. Therefore I added a keyword parameters `attriblocation` for those (just like the `fragdatalocation`) to the internal constructor of Program.
- some really minor things like a forgotten reference to the `context` field of the `Shader` in the implementation of the "=="-operator for `Shader`


> About it being clean, I feel like it is clean enough at least for what I use it :p, but I'm open to suggestions. From my point of view nothing really still needs to be done to get the core functionality working, but I for sure don't use all the features of OpenGL so your mileage may vary.

That is also good to hear, it means I can start replacing my own code step by step with the abstractions from your branch. Since I will write tests for every functionality i need anyway i can offer to add tests to the branch (by PRs or however you prefer it). The tests currently implemented in the tests folder do not run, right?

**Comment by louisponet** [comment]
You're right on the gluniform with matrices. I usually only use `StaticArrays` in my uniforms, and I didn't want to make `GLAbstraction` depend on anything except for `ModernGL` (tough, but I think it's a good idea), so I guess I never got around to implement a gluniform for generic arrays. With regards to the `StaticArrays/GeometryTypes` I had these lines in [extensions.jl](https://github.com/louisponet/Glimpse.jl/blob/6f77aca6c47d7857e267b644defb47bb194a19cf/src/extensions.jl#L70).

I've never got around to actually rewrite the tests, it would be most appreciated if you could make PR's with yours, same goes for the other fixes you have! Any help is very welcome. 

I think there is also something going wrong with how I handle setting texture uniforms, but I need to look into that.

@SimonDanisch Since you have what you use of `GLAbstraction` incorporated in `GLMakie`, should I make a PR with the v1.0 + refactor branch I've been working on so far, and make it the new master? I don't think any package except for ours depends on `GLAbstraction` do they?

**Comment by SimonDanisch** [comment]
Yeah, I don't see any reason why you shouldn't start owning GLAbstraction, if you use it and ported it to 1.0 ;)

**Pull request merged**